libraries/Microsoft.Bot.Builder.AI.QnA/QnAMakerAction.cs

```diff
@@ -0,0 +1,162 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.Dialogs;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Builder.AI.QnA
+{
+    /// <summary>
+    /// QnAMaker dialog which uses QnAMaker to get an answer.
+    /// </summary>
+    public class QnAMakerAction : Dialog
+    {
+        private QnAMaker qnamaker;
+        private readonly HttpClient httpClient;
+
+        public QnAMakerAction(
+            string knowledgeBaseId, 
+            string endpointKey, 
+            string hostName, 
+            string noAnswer = QnAMakerActionBuilder.DefaultNoAnswer, 
+            float threshold = QnAMakerActionBuilder.DefaultThreshold, 
+            string activeLearningCardTitle = QnAMakerActionBuilder.DefaultCardTitle, 
+            string cardNoMatchText = QnAMakerActionBuilder.DefaultCardNoMatchText, 
+            string cardNoMatchResponse = QnAMakerActionBuilder.DefaultCardNoMatchResponse, 
+            Metadata[] strictFilters = null,  
+            HttpClient httpClient = null, 
+            [CallerFilePath] string sourceFilePath = "", 
+            [CallerLineNumber] int sourceLineNumber = 0
+            )
+            : base()
+        {
+            this.RegisterSourceLocation(sourceFilePath, sourceLineNumber);
+            this.KnowledgeBaseId = knowledgeBaseId ?? throw new ArgumentNullException(nameof(knowledgeBaseId));
+            this.HostName = hostName ?? throw new ArgumentNullException(nameof(hostName));
+            this.EndpointKey = endpointKey ?? throw new ArgumentNullException(nameof(endpointKey));
+            this.Threshold = threshold;
+            this.NoAnswer = noAnswer;
+            this.ActiveLearningCardTitle = activeLearningCardTitle;
+            this.CardNoMatchText = cardNoMatchText;
+            this.CardNoMatchResponse = cardNoMatchResponse;
+            this.StrictFilters = strictFilters;
+            this.httpClient = httpClient;
+        }
+
+        [JsonConstructor]
+        public QnAMakerAction([CallerFilePath] string sourceFilePath = "", [CallerLineNumber] int sourceLineNumber = 0)
+            : base()
+        {
+            this.RegisterSourceLocation(sourceFilePath, sourceLineNumber);
+        }
+
+        [JsonProperty("knowledgeBaseId")]
+        public string KnowledgeBaseId { get; set; }
+
+        [JsonProperty("hostname")]
+        public string HostName { get; set; }
+
+        [JsonProperty("endpointKey")]
+        public string EndpointKey { get; set; }
+
+        [JsonProperty("threshold")]
+        public float Threshold { get; set; }
+
+        [JsonProperty("noAnswer")]
+        public string NoAnswer { get; set; }
```

Comment:
@vishwacsena  and I just discussed this a bit (and I had mentioned it to you in the past as well). Instead of these strings, these could be ITemplate<Activity> so that 1) text is resolved through our LG system and I can specify morethan just the text but the entire activity.

See here for an example of an activity template instead of a string:

SendActivity has an ITemplate<Activity> that defines the template here: https://github.com/microsoft/botbuilder-dotnet/blob/4.Future/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/SendActivity.cs#L39

And then it is evaluated here: https://github.com/microsoft/botbuilder-dotnet/blob/4.Future/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/SendActivity.cs#L48

Reply:
Thanks for the docs. Implemented the changes.